README.md
[![npm version](https://badge.fury.io/js/%40taimos%2Fcdk-construct-hosting.svg)](https://badge.fury.io/js/%40taimos%2Fcdk-construct-hosting)
[![PyPI version](https://badge.fury.io/py/taimos.cdk-construct-hosting.svg)](https://badge.fury.io/py/taimos.cdk-construct-hosting)


# A CDK L3 Construct for web hosting

## Installation

You can install the library into your project using npm or pip.

```bash
npm install @taimos/cdk-construct-hosting

pip3 install taimos.cdk-construct-hosting
```


# Contributing

We welcome community contributions and pull requests. 

# License

The CDK construct library is distributed under the [Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0).

See [LICENSE](./LICENSE) for more information.

CHANGELOG.md
# [1.1.0](https://github.com/taimos/cdk-construct-hosting/compare/v1.0.0...v1.1.0) (2020-03-19)


### Bug Fixes

* build scripts ([c37f2ed](https://github.com/taimos/cdk-construct-hosting/commit/c37f2ed63c5e5bd114fe83d605568516c0bfc92c))
* Repo URL in package.json corrected ([797de73](https://github.com/taimos/cdk-construct-hosting/commit/797de7342a4611f647c238ad1d376206c064d5ac))


### Features

* build on PR ([0d72fc6](https://github.com/taimos/cdk-construct-hosting/commit/0d72fc6787265cec7b6242eaf0871f78254def66))
* update to CDK 1.30 ([1c3fcd4](https://github.com/taimos/cdk-construct-hosting/commit/1c3fcd4ddfcbe48c57c9746b15e4261a1551df17))
* Upgrade to CDK 1.21.0 ([166696e](https://github.com/taimos/cdk-construct-hosting/commit/166696e942c274cec853f54f068d8121866bd3a7))
* upgrade to CDK 1.26.0 ([1c35ec7](https://github.com/taimos/cdk-construct-hosting/commit/1c35ec7bfe52a1d8c701a53cd903000d82285a91))



# [1.0.0](https://github.com/taimos/cdk-construct-hosting/compare/ce384db08fe1daa37811ddf6ed0c5b92565fd911...v1.0.0) (2020-01-12)

### Features

* first initial version


tst/index.test.ts
import { SinglePageAppHosting } from '../lib/index';
import { App, Stack } from '@aws-cdk/core';
import { expect as expectCDK, haveResourceLike } from '@aws-cdk/assert';

test('should be valid', () => {
    const mockApp = new App();
    const stack = new Stack(mockApp, 'testing-stack');
    new SinglePageAppHosting(stack, 'SPA', {
        domainName: 'example.net',
        certArn: 'arn:aws:acm:us-east-1:123456789012:certificate/foobar',
        zoneId: '1234567890',
    });

    expectCDK(stack).to(haveResourceLike('AWS::CloudFront::Distribution', {
        "DistributionConfig": {
            "Aliases": [
                "example.net"
            ],
            "Comment": "example.net Website",
            "CustomErrorResponses": [
                {
                    "ErrorCode": 403,
                    "ResponseCode": 200,
                    "ResponsePagePath": "/index.html"
                },
                {
                    "ErrorCode": 404,
                    "ResponseCode": 200,
                    "ResponsePagePath": "/index.html"
                }
            ],
            "DefaultRootObject": "index.html",
            "ViewerCertificate": {
                "AcmCertificateArn": "arn:aws:acm:us-east-1:123456789012:certificate/foobar",
                "SslSupportMethod": "sni-only"
            }
        }
    }));
    // console.log(JSON.stringify(SynthUtils.toCloudFormation(stack, {}), null, 2));
});

test('should auto generate certificates', () => {
    const mockApp = new App();
    const stack = new Stack(mockApp, 'testing-stack');
    new SinglePageAppHosting(stack, 'SPA', {
        domainName: 'example.net',
        zoneId: '1234567890',
    });

    expectCDK(stack).to(haveResourceLike('AWS::CloudFormation::CustomResource', {
        "DomainName": "example.net",
        "HostedZoneId": "1234567890",
        "Region": "us-east-1"
    }));
    // console.log(JSON.stringify(SynthUtils.toCloudFormation(stack, {}), null, 2));
});


lib/index.ts
import { Certificate, DnsValidatedCertificate } from '@aws-cdk/aws-certificatemanager';
import { CloudFrontWebDistribution, OriginAccessIdentity, PriceClass, ViewerCertificate, ViewerProtocolPolicy } from '@aws-cdk/aws-cloudfront';
import { ARecord, HostedZone, RecordTarget } from '@aws-cdk/aws-route53';
import { CloudFrontTarget } from '@aws-cdk/aws-route53-targets';
import { Bucket } from '@aws-cdk/aws-s3';
import { BucketDeployment, Source } from '@aws-cdk/aws-s3-deployment';
import { Construct } from '@aws-cdk/core';

export interface SinglePageAppHostingProps {
    /**
     * Name of the domain to deploy
     */
    readonly domainName : string;
    /**
     * Name of the HostedZone of the domain
     * 
     * @default - same a the domain name
     */
    readonly zoneName? : string;
    /**
     * Name of the index document
     * 
     * @default index.html
     */
    readonly indexFile? : string;
    /**
     * The ARN of the certificate; Has to be in us-east-1
     *
     * @default - create a new certificate in us-east-1
     */
    readonly certArn? : string;
    /**
     * ID of the HostedZone of the domain
     *
     * @default - lookup zone from context using the zone name
     */
    readonly zoneId? : string;
    /**
     * local folder with contents for the website bucket
     *
     * @default - no file deployment
     */
    readonly webFolder? : string;
}

export class SinglePageAppHosting extends Construct {

    public readonly webBucket : Bucket;

    public readonly distribution : CloudFrontWebDistribution;

    constructor(scope : Construct, id : string, props : SinglePageAppHostingProps) {
        super(scope, id);

        const zoneName = props.zoneName ?? props.domainName;

        const zone = props.zoneId ? HostedZone.fromHostedZoneAttributes(this, 'HostedZone', {
            hostedZoneId: props.zoneId,
            zoneName,
        }) : HostedZone.fromLookup(this, 'HostedZone', { domainName: zoneName });

        const cert = props.certArn ?
            Certificate.fromCertificateArn(this, 'Certificate', props.certArn) :
            new DnsValidatedCertificate(this, 'Certificate', {
                hostedZone: zone,
                domainName: props.domainName,
                region: 'us-east-1',
            });

        const originAccessIdentity = new OriginAccessIdentity(this, 'OAI', { comment: props.domainName });
        this.webBucket = new Bucket(this, 'WebBucket', { websiteIndexDocument: props.indexFile ?? 'index.html' });
        this.webBucket.grantRead(originAccessIdentity);

        this.distribution = new CloudFrontWebDistribution(this, 'Distribution', {
            originConfigs: [{
                behaviors: [{ isDefaultBehavior: true }],
                s3OriginSource: {
                    s3BucketSource: this.webBucket,
                    originAccessIdentity,
                },
            }],
            viewerCertificate: ViewerCertificate.fromAcmCertificate(cert, {
                aliases: [props.domainName],
            }),
            errorConfigurations: [{
                errorCode: 403,
                responseCode: 200,
                responsePagePath: `/${props.indexFile ?? 'index.html'}`,
            }, {
                errorCode: 404,
                responseCode: 200,
                responsePagePath: `/${props.indexFile ?? 'index.html'}`,
            }],
            comment: `${props.domainName} Website`,
            priceClass: PriceClass.PRICE_CLASS_ALL,
            viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
        });

        if (props.webFolder) {
            new BucketDeployment(this, 'DeployWebsite', {
                sources: [Source.asset(props.webFolder)],
                destinationBucket: this.webBucket,
                distribution: this.distribution,
            });
        }

        new ARecord(this, 'AliasRecord', {
            recordName: props.domainName,
            zone,
            target: RecordTarget.fromAlias(new CloudFrontTarget(this.distribution)),
        });
    }
}
